tracing/smithy-otel-tracing/adapt.go

```diff
@@ -0,0 +1,186 @@
+	}
+
+	var options tracing.SpanOptions
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	kind := toOTELSpanKind(options.Kind)
+	ctx, theirs := t.otel.Start(ctx, name, oteltrace.WithSpanKind(kind))
+
+	ours = &span{
+		otel: theirs,
+		name: name,
+	}
+	for k, v := range options.Properties.Values() {
+		ours.SetProperty(k, v)
+	}
+	return tracing.WithSpan(ctx, ours) /* (1) */, ours
+}
+
+type span struct {
+	otel oteltrace.Span
+	name string
+}
+
+var _ tracing.Span = (*span)(nil)
+
+func (s *span) Name() string {
+	return s.name
+}
+
+func (s *span) Context() tracing.SpanContext {
+	ctx := s.otel.SpanContext()
+	return tracing.SpanContext{
+		TraceID:  ctx.TraceID().String(),
+		SpanID:   ctx.SpanID().String(),
+		IsRemote: ctx.IsRemote(),
+	}
+}
+
+func (s *span) AddEvent(name string, opts ...tracing.EventOption) {
+	var options tracing.EventOptions
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	s.otel.AddEvent(name, oteltrace.WithAttributes(
+		toOTELKeyValues(options.Properties)...,
+	))
+}
+
+func (s *span) SetProperty(k, v any) {
+	if kv, ok := toOTELKeyValue(k, v); ok {
+		s.otel.SetAttributes(kv)
+	}
+}
+
+func (s *span) SetStatus(status tracing.SpanStatus) {
+	s.otel.SetStatus(toOTELSpanStatus(status), "")
+}
+
+func (s *span) End() {
+	s.otel.End()
+}
+
+func toOTELSpanKind(v tracing.SpanKind) oteltrace.SpanKind {
+	switch v {
+	case tracing.SpanKindClient:
+		return oteltrace.SpanKindClient
+	case tracing.SpanKindServer:
+		return oteltrace.SpanKindServer
+	case tracing.SpanKindProducer:
+		return oteltrace.SpanKindProducer
+	case tracing.SpanKindConsumer:
+		return oteltrace.SpanKindConsumer
+	default:
+		return oteltrace.SpanKindInternal
+	}
+}
+
+func toOTELSpanStatus(v tracing.SpanStatus) otelcodes.Code {
+	switch v {
+	case tracing.SpanStatusOK:
+		return otelcodes.Ok
+	case tracing.SpanStatusError:
+		return otelcodes.Error
+	default:
+		return otelcodes.Unset
+	}
+}
+
+func toOTELKeyValue(k, v any) (otelattribute.KeyValue, bool) {
+	kk, ok := k.(string)
+	if !ok {
+		return otelattribute.KeyValue{}, false
+	}
+
+	switch vv := v.(type) {
+	case bool:
+		return otelattribute.Bool(kk, vv), true
+	case []bool:
+		return otelattribute.BoolSlice(kk, vv), true
+	case int:
+		return otelattribute.Int(kk, vv), true
+	case []int:
+		return otelattribute.IntSlice(kk, vv), true
+	case int64:
+		return otelattribute.Int64(kk, vv), true
+	case []int64:
+		return otelattribute.Int64Slice(kk, vv), true
+	case float64:
+		return otelattribute.Float64(kk, vv), true
+	case []float64:
+		return otelattribute.Float64Slice(kk, vv), true
+	case string:
+		return otelattribute.String(kk, vv), true
+	case []string:
+		return otelattribute.StringSlice(kk, vv), true
+	default:
+		return otelattribute.KeyValue{}, false
+	}
+}
+
+func toOTELKeyValues(props smithy.Properties) []otelattribute.KeyValue {
+	var kvs []otelattribute.KeyValue
+	for k, v := range props.Values() {
+		if kv, ok := toOTELKeyValue(k, v); ok {
```

Comment:
I see that you mention that all of these values will be silently ignored, which can make this very hard to debug if things go wrong. Should we at least log this?

Reply:
Actually yes, couldn't it just be a span event?

Reply:
Discussed offline, it's probably better to just map to a string representation as default. Pushed.